Kudu.Core/Deployment/StructuredTextDocument.cs

```diff
@@ -0,0 +1,119 @@
+﻿using Kudu.Core.Infrastructure;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Kudu.Core.Deployment
+{
+    public class StructuredTextDocument<T>
+    {
+        private readonly static object DocumentLock = new object();
+        private const string LogTemplate = "{0}{1}";
+        private readonly string _path;
+        private readonly Func<T, string> _serializer;
+        private readonly Func<string, T> _deserializer;
+
+        public StructuredTextDocument(string path, Func<T, string> serializer, Func<string, T> deserializer)
+        {
+            _path = path;
+            _serializer = serializer;
+            _deserializer = deserializer;
+        }
+
+        public IEnumerable<StructuredTextDocumentEntry<T>> GetDocument()
+        {
+            return ParseLines(FileSystemHelpers.ReadAllLines(_path), startIndex: 0, depth: 0);
+        }
+
+        public void Write(T value, int depth)
```

Comment:
is the "Write" function a public API method? what does "depth" mean?

will be good to have more comment and class document to describe how it work, and have some sample output in the class comment would be a great help when reading the code.